### PR TITLE
feat: asset registry refactored

### PR DESCRIPTION
**Pull request by enthusiastmartin**
New pallet-registry added.

It is a rework of original and very simple asset-registry pallet. This new implementation attempts to be more robust and easier to use. 

New Features compared to previous implementation includes: 
- dispatchables  for registering and updating assets
- string limit for asset name
- native locations ( use for xcm ) - it possible to store an origin location of an asset. This can be used in xcm currency/location conversions ( no need to specify/hard-code it in runtime)
- asset detail and metadata
- better handling of 'next asset id' - our asset registry uses sequential ids which is assigned for each newly registered asset.

### CODE REVIEW

File: pallets/registry/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,406 @@
...
+			})
+		}
+
+		/// Set metadata for an asset.
+		///
+		/// - `asset_id`: Asset identifier.
+		/// - `symbol`: The exchange symbol for this asset. Limited in length by `StringLimit`.
+		/// - `decimals`: The number of decimals this asset uses to represent one unit.
+		///
+		/// Emits `MetedataSet` event when successful.
```
Comment (green-jay) [inline pallets/registry/src/lib.rs:283]:
```suggestion
		/// Emits `MetadataSet` event when successful.
```

File: pallets/registry/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,406 @@
...
+	}
+}
+
+impl<T: Config> Pallet<T> {
+	/// Convert Vec<u8> to BoundedVec so it respects the max set limit, otherwise return BadMetadata error
+	fn to_bounded_name(name: Vec<u8>) -> Result<BoundedVec<u8, T::StringLimit>, Error<T>> {
+		name.try_into().map_err(|_| Error::<T>::BadMetadata)
+	}
+
+	///Register new asset.
```
Comment (green-jay) [inline pallets/registry/src/lib.rs:344]:
```suggestion
	/// Register new asset.
```

File: pallets/registry/src/weights.rs [hunk 3]
```diff
@@ -0,0 +1,93 @@
+// This file is part of HydraDX-node.
```
Comment (green-jay) [inline pallets/registry/src/weights.rs:1]: what is the purpose of this file? is it still WIP?

File: pallets/registry/README.md [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+- **AssetIds** - list of existing asset ids
+- **AssetDetail** - details of an asset such as type, name or whether it is locked or not.
+- **AssetMetadata** - additional optional metadata of an asset ( symbol, decimals)
+- **AssetLocation** - information of native location of an asset. Used in XCM.
+
+### Implementation detail
+
+For each newly registered asset, a sequential id is assigned to that asset. This id identifies the asset and can be used directly in transfers or any other operation which works with an asset ( without performing any additioanl asset check or asset retrieval).
+
+There is a mapping between the name aand asset id stored as well which helps and is used in AMM Implementation where there is a need to register a pool asset and only name is provided ( see `get_or_create_asset` ).
```
Comment (jak-pan) [inline pallets/registry/README.md:19]:
```suggestion
There is a mapping between the name and asset id stored as well, which helps and is used in AMM Implementation where there is a need to register a pool asset and only name is provided ( see `get_or_create_asset` ).
```

File: pallets/registry/src/weights.rs (hunk 3 above)
Comment (enthusiastmartin) [reply pallets/registry/src/weights.rs:1]: benchmarks in progress. It will be completed today and all pallets ( not only registry) will be rebenchmarked

Comment (green-jay) [reply pallets/registry/src/weights.rs:1]: So far PR looks good, just a few cosmetic changes. Will get back to it next week then.

File: pallets/registry/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,407 @@
...
+
+					// update also name map - remove old one first
+					AssetIds::<T>::remove(&detail.name);
+					AssetIds::<T>::insert(&bn, asset_id);
+				}
+
+				detail.name = bn.clone();
+				detail.asset_type = asset_type;
+
+				Self::deposit_event(Event::Updated(asset_id, bn, asset_type));
```
Comment (jak-pan) [inline pallets/registry/src/lib.rs:272]: rename bn pls

**Review by lumir-mrkva — APPROVED** [review]
LGTM

**Pull request merged**